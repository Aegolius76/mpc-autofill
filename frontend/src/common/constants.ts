export const ProjectName = "MPC Autofill";

import { CardType, Faces } from "@/common/types";

export const Card: CardType = "CARD";
export const Cardback: CardType = "CARDBACK";
export const Token: CardType = "TOKEN";

export const CardTypeSeparator = ":";

export const CardTypePrefixes: { [prefix: string]: CardType } = {
  "": Card,
  b: Cardback,
  t: Token,
};

export const ReversedCardTypePrefixes = Object.fromEntries(
  Object.keys(CardTypePrefixes).map((prefix: string) => [
    CardTypePrefixes[prefix],
    prefix.length > 0 ? prefix + CardTypeSeparator : prefix,
  ])
);

export const Front: Faces = "front";
export const Back: Faces = "back";
export const FaceSeparator = "|";

export const ToggleButtonHeight = 38; // pixels

export const MinimumDPI = 0;
export const MaximumDPI = 1500;
export const DPIStep = 50;

export const MaximumSize = 30; // megabytes
export const SizeStep = 1;

export const CSRFKey = "csrftoken";
export const SearchSettingsKey = "searchSettings";
export const GoogleAnalyticsConsentKey = "googleAnalyticsConsent";
export const BackendURLKey = "backendURL";

export const Brackets: Array<number> = [
  18, 36, 55, 72, 90, 108, 126, 144, 162, 180, 198, 216, 234, 396, 504, 612,
];

export const ProjectMaxSize: number = Brackets[Brackets.length - 1];

export enum QueryTags {
  BackendSpecific = "backendSpecific",
}
