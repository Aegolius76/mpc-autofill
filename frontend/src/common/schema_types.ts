/* eslint-disable */
/**
 * This file was automatically generated by json-schema-to-typescript.
 * DO NOT MODIFY IT BY HAND. Instead, modify the source JSONSchema file,
 * and run json-schema-to-typescript to regenerate this file.
 */

export interface SearchQuery {
  query: string | null;
  card_type: "CARD" | "CARDBACK" | "TOKEN";
}
/* eslint-disable */
/**
 * This file was automatically generated by json-schema-to-typescript.
 * DO NOT MODIFY IT BY HAND. Instead, modify the source JSONSchema file,
 * and run json-schema-to-typescript to regenerate this file.
 */

/**
 * @minItems 2
 * @maxItems 2
 */
export type SourceRow = [number, boolean];

export interface SearchSettings {
  searchTypeSettings: SearchTypeSettings;
  sourceSettings: SourceSettings;
  filterSettings: FilterSettings;
}
export interface SearchTypeSettings {
  /**
   * Whether fuzzy search is enabled
   */
  fuzzySearch: boolean;
  /**
   * Whether search settings apply to cardbacks or not
   */
  filterCardbacks: boolean;
}
export interface SourceSettings {
  /**
   * The list of sources in the order they should be searched
   */
  sources: SourceRow[] | null;
}
export interface FilterSettings {
  /**
   * The minimum DPI that cards must meet to be included in search results
   */
  minimumDPI: number;
  /**
   * The maximum DPI that cards can have to be included in search results
   */
  maximumDPI: number;
  /**
   * The maximum filesize that cards can have to be included in search results
   */
  maximumSize: number;
}
