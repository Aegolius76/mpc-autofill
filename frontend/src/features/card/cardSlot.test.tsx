import { fireEvent, screen, waitFor } from "@testing-library/react";

import App from "@/app/app";
import { Back, Card, Front } from "@/common/constants";
import {
  cardDocument1,
  cardDocument2,
  cardDocument3,
  localBackend,
  projectSelectedImage1,
  projectSelectedImage2,
  projectThreeMembersSelectedImage1,
} from "@/common/test-constants";
import {
  expectCardbackSlotState,
  expectCardGridSlotState,
  expectCardSlotToNotExist,
  openCardSlotGridSelector,
  renderWithProviders,
  selectSlot,
} from "@/common/test-utils";
import {
  cardbacksTwoResults,
  cardDocumentsFourResults,
  cardDocumentsOneResult,
  cardDocumentsThreeResults,
  defaultHandlers,
  searchResultsFourResults,
  searchResultsOneResult,
  searchResultsThreeResults,
  sourceDocumentsOneResult,
  sourceDocumentsThreeResults,
} from "@/mocks/handlers";
import { server } from "@/mocks/server";

//# region snapshot tests

test("the html structure of a CardSlot with a single search result, no image selected", async () => {
  server.use(
    cardDocumentsOneResult,
    sourceDocumentsOneResult,
    searchResultsOneResult
  );
  renderWithProviders(<App />, {
    preloadedState: {
      backend: localBackend,
      project: {
        members: [
          {
            front: {
              query: { query: "my search query", card_type: Card },
              selectedImage: null,
            },
            back: null,
          },
        ],
        cardback: null,
      },
    },
  });
  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 1);
  expect(screen.getByTestId("front-slot0")).toMatchSnapshot();
});

test("the html structure of a CardSlot with a single search result, slot selected", async () => {
  server.use(
    cardDocumentsOneResult,
    sourceDocumentsOneResult,
    searchResultsOneResult
  );
  renderWithProviders(<App />, {
    preloadedState: {
      backend: localBackend,
      project: {
        members: [
          {
            front: {
              query: { query: "my search query", card_type: Card },
              selectedImage: null,
            },
            back: null,
          },
        ],
        cardback: null,
      },
    },
  });
  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 1);
  screen.getByLabelText("select-front0").click();
  await waitFor(() =>
    expect(screen.getByLabelText("select-front0").firstChild).toHaveClass(
      "bi-check-square"
    )
  );
  expect(screen.getByTestId("front-slot0")).toMatchSnapshot();
});

test("the html structure of a CardSlot with a single search result, image selected", async () => {
  server.use(
    cardDocumentsOneResult,
    sourceDocumentsOneResult,
    searchResultsOneResult
  );

  renderWithProviders(<App />, {
    preloadedState: {
      backend: localBackend,
      project: projectSelectedImage1,
    },
  });

  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 1);
  expect(screen.getByTestId("front-slot0")).toMatchSnapshot();
});

test("the html structure of a CardSlot with multiple search results, image selected", async () => {
  server.use(
    cardDocumentsThreeResults,
    sourceDocumentsOneResult,
    searchResultsThreeResults
  );

  renderWithProviders(<App />, {
    preloadedState: { backend: localBackend, project: projectSelectedImage1 },
  });

  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 3);
  expect(screen.getByTestId("front-slot0")).toMatchSnapshot();
});

test("the html structure of a CardSlot's grid selector, cards faceted by source", async () => {
  server.use(
    cardDocumentsFourResults,
    sourceDocumentsThreeResults,
    searchResultsFourResults
  );

  renderWithProviders(<App />, {
    preloadedState: {
      backend: localBackend,
      project: projectSelectedImage1,
    },
  });

  const gridSelector = await openCardSlotGridSelector(1, Front, 1, 4);
  expect(gridSelector).toMatchSnapshot();
});

test("the html structure of a CardSlot's grid selector, cards grouped together", async () => {
  server.use(
    cardDocumentsFourResults,
    sourceDocumentsThreeResults,
    searchResultsFourResults
  );

  renderWithProviders(<App />, {
    preloadedState: {
      backend: localBackend,
      project: projectSelectedImage1,
      viewSettings: {
        frontsVisible: true,
        facetBySource: false,
        sourcesVisible: {},
      },
    },
  });

  const gridSelector = await openCardSlotGridSelector(1, Front, 1, 4);
  expect(gridSelector).toMatchSnapshot();
});

//# endregion

test("switching to the next image in a CardSlot", async () => {
  server.use(
    cardDocumentsThreeResults,
    sourceDocumentsOneResult,
    searchResultsThreeResults,
    ...defaultHandlers
  );
  renderWithProviders(<App />, {
    preloadedState: {
      backend: localBackend,
      project: projectSelectedImage1,
    },
  });

  // from preloaded state
  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 3);

  screen.getByText("❯").click();

  await expectCardGridSlotState(1, Front, cardDocument2.name, 2, 3);
});

test("switching to the previous image in a CardSlot", async () => {
  server.use(
    cardDocumentsThreeResults,
    sourceDocumentsOneResult,
    searchResultsThreeResults,
    ...defaultHandlers
  );
  renderWithProviders(<App />, {
    preloadedState: {
      backend: localBackend,
      project: projectSelectedImage2,
    },
  });

  // from preloaded state
  await expectCardGridSlotState(1, Front, cardDocument2.name, 2, 3);

  screen.getByText("❮").click();

  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 3);
});

test("switching images in a CardSlot wraps around", async () => {
  server.use(
    cardDocumentsThreeResults,
    sourceDocumentsOneResult,
    searchResultsThreeResults,
    ...defaultHandlers
  );
  renderWithProviders(<App />, {
    preloadedState: {
      backend: localBackend,
      project: projectSelectedImage2,
    },
  });

  // from preloaded state
  await expectCardGridSlotState(1, Front, cardDocument2.name, 2, 3);

  // page between images and ensure that wrapping around works
  screen.getByText("❯").click();
  await expectCardGridSlotState(1, Front, cardDocument3.name, 3, 3);

  screen.getByText("❯").click();
  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 3);

  screen.getByText("❮").click();
  await expectCardGridSlotState(1, Front, cardDocument3.name, 3, 3);
});

test("selecting an image in a CardSlot via the grid selector", async () => {
  server.use(
    cardDocumentsThreeResults,
    sourceDocumentsOneResult,
    searchResultsThreeResults,
    ...defaultHandlers
  );
  renderWithProviders(<App />, {
    preloadedState: {
      backend: localBackend,
      project: projectSelectedImage2,
    },
  });

  // from preloaded state
  await expectCardGridSlotState(1, Front, cardDocument2.name, 2, 3);

  await waitFor(() => screen.getByText("2 / 3").click());
  await waitFor(() => expect(screen.getByText("Select Version")));

  expect(screen.getByText("Option 2")).toBeInTheDocument();
  expect(screen.getByText("Option 3")).toBeInTheDocument();
  screen.getByText("Option 1").click();

  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 3);
});

test("deleting a CardSlot", async () => {
  server.use(
    cardDocumentsThreeResults,
    sourceDocumentsOneResult,
    searchResultsThreeResults,
    ...defaultHandlers
  );
  renderWithProviders(<App />, {
    preloadedState: {
      backend: localBackend,
      project: projectSelectedImage1,
    },
  });
  await waitFor(() =>
    expect(screen.getByText(cardDocument1.name)).toBeInTheDocument()
  );

  screen.getByLabelText("remove-front0").click();

  await expectCardSlotToNotExist(1);
});

test("deleting multiple CardSlots", async () => {
  server.use(
    cardDocumentsThreeResults,
    sourceDocumentsOneResult,
    searchResultsThreeResults,
    ...defaultHandlers
  );
  renderWithProviders(<App />, {
    preloadedState: {
      backend: localBackend,
      project: projectThreeMembersSelectedImage1,
    },
  });
  await waitFor(() =>
    expect(screen.getAllByText(cardDocument1.name)).toHaveLength(3)
  );

  screen.getByLabelText("remove-front0").click();
  screen.getByLabelText("remove-front1").click();

  await expectCardSlotToNotExist(2);
  await expectCardSlotToNotExist(3);
});

test("CardSlot automatically selects the first search result", async () => {
  server.use(
    cardDocumentsThreeResults,
    sourceDocumentsOneResult,
    searchResultsThreeResults,
    ...defaultHandlers
  );
  renderWithProviders(<App />, {
    preloadedState: {
      backend: localBackend,
      project: {
        members: [
          {
            front: {
              query: { query: "my search query", card_type: Card },
              selectedImage: null,
            },
            back: null,
          },
        ],
        cardback: null,
      },
    },
  });

  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 3);
});

test("CardSlot automatically deselects invalid image then selects the first search result", async () => {
  server.use(
    cardDocumentsThreeResults,
    sourceDocumentsOneResult,
    searchResultsOneResult,
    ...defaultHandlers
  );
  renderWithProviders(<App />, {
    preloadedState: {
      backend: localBackend,
      project: projectSelectedImage2, // not in search results
    },
  });

  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 1);
});

test("CardSlot uses cardbacks as search results for backs with no search query", async () => {
  server.use(
    cardDocumentsThreeResults,
    cardbacksTwoResults,
    sourceDocumentsOneResult,
    ...defaultHandlers
  );
  renderWithProviders(<App />, {
    preloadedState: {
      backend: localBackend,
      project: {
        members: [
          {
            front: null,
            back: {
              query: null,
              selectedImage: cardDocument1.identifier,
            },
          },
        ],
        cardback: null,
      },
    },
  });

  await expectCardGridSlotState(1, Back, cardDocument1.name, 1, 2);
  await expectCardbackSlotState(cardDocument1.name, 1, 2);
});

test("CardSlot defaults to project cardback for backs with no search query", async () => {
  server.use(
    cardDocumentsThreeResults,
    cardbacksTwoResults,
    sourceDocumentsOneResult,
    ...defaultHandlers
  );
  renderWithProviders(<App />, {
    preloadedState: {
      backend: localBackend,
      project: {
        members: [
          {
            front: null,
            back: null,
          },
        ],
        cardback: cardDocument1.identifier,
      },
    },
  });

  await expectCardGridSlotState(1, Back, cardDocument1.name, 1, 2);
  await expectCardbackSlotState(cardDocument1.name, 1, 2);
});

test("double clicking the select button selects all slots for the same query", async () => {
  server.use(
    cardDocumentsOneResult,
    sourceDocumentsOneResult,
    searchResultsOneResult
  );
  renderWithProviders(<App />, {
    preloadedState: {
      backend: localBackend,
      project: {
        members: [
          {
            front: {
              query: { query: "my search query", card_type: Card },
              selectedImage: null,
            },
            back: null,
          },
          {
            front: {
              query: { query: "my search query", card_type: Card },
              selectedImage: null,
            },
            back: null,
          },
        ],
        cardback: null,
      },
    },
  });
  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 1);
  await expectCardGridSlotState(2, Front, cardDocument1.name, 1, 1);

  await selectSlot(1, Front);
  await selectSlot(1, Front, 2);
  await waitFor(() =>
    expect(screen.getByLabelText("select-front0").firstChild).toHaveClass(
      "bi-check-square"
    )
  );
  await waitFor(() =>
    expect(screen.getByLabelText("select-front1").firstChild).toHaveClass(
      "bi-check-square"
    )
  );
});
