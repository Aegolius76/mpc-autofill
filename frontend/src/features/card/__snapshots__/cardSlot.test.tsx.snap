// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the html structure of a CardSlot with a single search result, image selected 1`] = `
<div
  data-testid="front-slot0"
>
  <div
    class="mpccard mpccard-hover card"
  >
    <div
      class="pb-0 text-center card-header"
    >
      <p
        class="mpccard-slot"
      >
        Slot 1
      </p>
      <button
        class="card-select"
      >
        <i
          class="bi bi-square"
        />
      </button>
      <button
        class="remove"
      >
        <i
          aria-label="remove-front0"
          class="bi bi-x-circle"
        />
      </button>
    </div>
    <div>
      <div
        class="rounded-lg shadow-lg ratio ratio-7x5"
        style="z-index: 0;"
      >
        <img
          alt="Card 1"
          class="card-img card-img-fade-in"
          data-nimg="fill"
          decoding="async"
          loading="lazy"
          src=""
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent; z-index: 1; opacity: 1;"
        />
      </div>
      <div
        class="mb-0 text-center card-body"
      >
        <div
          class="mpccard-name card-subtitle h6"
        >
          Card 1
        </div>
        <div
          class="mpccard-spacing"
        >
          <p
            class="mpccard-source card-text"
          >
            Source 1 [1200 DPI]
          </p>
        </div>
      </div>
    </div>
    <div
      class="padding-top card-footer"
      style="padding-top: 50px;"
    >
      <p
        class="mpccard-counter text-center align-middle"
      >
        1 / 
        1
      </p>
    </div>
  </div>
</div>
`;

exports[`the html structure of a CardSlot with a single search result, no image selected 1`] = `
<div
  data-testid="front-slot0"
>
  <div
    class="mpccard mpccard-hover card"
  >
    <div
      class="pb-0 text-center card-header"
    >
      <p
        class="mpccard-slot"
      >
        Slot 1
      </p>
      <button
        class="card-select"
      >
        <i
          class="bi bi-square"
        />
      </button>
      <button
        class="remove"
      >
        <i
          aria-label="remove-front0"
          class="bi bi-x-circle"
        />
      </button>
    </div>
    <div>
      <div
        class="rounded-lg shadow-lg ratio ratio-7x5"
        style="z-index: 0;"
      >
        <img
          alt="Card 1"
          class="card-img card-img-fade-in"
          data-nimg="fill"
          decoding="async"
          loading="lazy"
          src=""
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent; z-index: 1; opacity: 1;"
        />
      </div>
      <div
        class="mb-0 text-center card-body"
      >
        <div
          class="mpccard-name card-subtitle h6"
        >
          Card 1
        </div>
        <div
          class="mpccard-spacing"
        >
          <p
            class="mpccard-source card-text"
          >
            Source 1 [1200 DPI]
          </p>
        </div>
      </div>
    </div>
    <div
      class="padding-top card-footer"
      style="padding-top: 50px;"
    >
      <p
        class="mpccard-counter text-center align-middle"
      >
        1 / 
        1
      </p>
    </div>
  </div>
</div>
`;

exports[`the html structure of a CardSlot with a single search result, slot selected 1`] = `
<div
  data-testid="front-slot0"
>
  <div
    class="mpccard mpccard-hover card"
  >
    <div
      class="pb-0 text-center card-header"
    >
      <p
        class="mpccard-slot"
      >
        Slot 1
      </p>
      <button
        aria-label="select-front0"
        class="card-select"
      >
        <i
          class="bi bi-check-square"
        />
      </button>
      <button
        class="remove"
      >
        <i
          aria-label="remove-front0"
          class="bi bi-x-circle"
        />
      </button>
    </div>
    <div>
      <div
        class="rounded-lg shadow-lg ratio ratio-7x5"
        style="z-index: 0;"
      >
        <img
          alt="Card 1"
          class="card-img card-img-fade-in"
          data-nimg="fill"
          decoding="async"
          loading="lazy"
          src=""
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent; z-index: 1; opacity: 1;"
        />
      </div>
      <div
        class="mb-0 text-center card-body"
      >
        <div
          class="mpccard-name card-subtitle h6"
        >
          Card 1
        </div>
        <div
          class="mpccard-spacing"
        >
          <p
            class="mpccard-source card-text"
          >
            Source 1 [1200 DPI]
          </p>
        </div>
      </div>
    </div>
    <div
      class="padding-top card-footer"
      style="padding-top: 50px;"
    >
      <p
        class="mpccard-counter text-center align-middle"
      >
        1 / 
        1
      </p>
    </div>
  </div>
</div>
`;

exports[`the html structure of a CardSlot with multiple search results, image selected 1`] = `
<div
  data-testid="front-slot0"
>
  <div
    class="mpccard mpccard-hover card"
  >
    <div
      class="pb-0 text-center card-header"
    >
      <p
        class="mpccard-slot"
      >
        Slot 1
      </p>
      <button
        class="card-select"
      >
        <i
          class="bi bi-square"
        />
      </button>
      <button
        class="remove"
      >
        <i
          aria-label="remove-front0"
          class="bi bi-x-circle"
        />
      </button>
    </div>
    <div>
      <div
        class="rounded-lg shadow-lg ratio ratio-7x5"
        style="z-index: 0;"
      >
        <img
          alt="Card 1"
          class="card-img card-img-fade-in"
          data-nimg="fill"
          decoding="async"
          loading="lazy"
          src=""
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent; z-index: 1; opacity: 1;"
        />
        <img
          alt="Card 2"
          class="sc-gueYoa jvIOZX card-img"
          data-nimg="fill"
          decoding="async"
          loading="lazy"
          src=""
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
        />
        <img
          alt="Card 3"
          class="sc-gueYoa jvIOZX card-img"
          data-nimg="fill"
          decoding="async"
          loading="lazy"
          src=""
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
        />
      </div>
      <div
        class="mb-0 text-center card-body"
      >
        <div
          class="mpccard-name card-subtitle h6"
        >
          Card 1
        </div>
        <div
          class="mpccard-spacing"
        >
          <p
            class="mpccard-source card-text"
          >
            Source 1 [1200 DPI]
          </p>
        </div>
      </div>
    </div>
    <div
      class="padding-top card-footer"
      style="padding-top: 50px;"
    >
      <button
        class="mpccard-counter-btn btn btn-outline-info"
        type="button"
      >
        1
         / 
        3
      </button>
      <div>
        <button
          class="prev btn btn-outline-primary"
          type="button"
        >
          ❮
        </button>
        <button
          class="next btn btn-outline-primary"
          type="button"
        >
          ❯
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`the html structure of a CardSlot's grid selector, cards faceted by source 1`] = `
<div
  class="modal-dialog modal-lg"
  data-testid="front-slot0-grid-selector"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="modal-title h4"
      >
        Select Version
      </div>
      <button
        aria-label="Close"
        class="btn-close"
        type="button"
      />
    </div>
    <div
      class="modal-body"
    >
      <div
        class="row"
      >
        <div
          class="col-md-8 col-sm-6"
        >
          <div
            class="btn toggle btn-md btn-success"
            role="button"
            style="width: 100%; height: 38px;"
          >
            <div
              class="toggle-group"
            >
              <span
                class="btn toggle-on btn-md flex-centre btn-success"
              >
                Facet By Source
              </span>
              <span
                class="btn toggle-off btn-md flex-centre btn-info"
              >
                Group All Cards Together
              </span>
              <span
                class="toggle-handle btn btn-md btn-default"
              />
            </div>
          </div>
        </div>
        <div
          class="col-md-4 col-sm-6"
        >
          <div
            class="d-grid gap-0"
          >
            <button
              class="btn btn-primary"
              type="button"
            >
              <i
                class="bi bi-arrows-collapse"
                style="padding-right: 0.5em;"
              />
               
              Collapse
               All
            </button>
          </div>
        </div>
      </div>
      <br />
      <div
        class="sticky-top"
        style="background-color: rgb(78, 93, 107); z-index: 100;"
      >
        <hr />
        <div
          class="d-flex ps-2 pe-2 hstack gap-2"
        >
          <h3
            class="orpheus prevent-select"
            style="font-style: italic;"
          >
            Source 1
          </h3>
          <h6
            class="text-primary prevent-select"
          >
            4
             version
            s
          </h6>
          <h4
            class="ms-auto bi bi-chevron-left rotate-90"
            data-testid="source_1-collapse-header"
            style="transition: all 0.25s 0s;"
          />
        </div>
      </div>
      <div
        class="collapse show"
        data-testid="source_1-collapse"
      >
        <hr />
        <div
          class="g-0 row row-cols-xxl-4 row-cols-xl-4 row-cols-lg-3 row-cols-md-2 row-cols-sm-2 row-cols-2"
        >
          <div
            class="mpccard mpccard-hover card"
          >
            <div
              class="pb-0 text-center card-header"
            >
              <p
                class="mpccard-slot"
              >
                Option 1
              </p>
            </div>
            <div>
              <div
                class="rounded-lg shadow-lg ratio ratio-7x5"
                style="z-index: 0;"
              >
                <img
                  alt="Card 1"
                  class="card-img card-img-fade-in"
                  data-nimg="fill"
                  decoding="async"
                  loading="lazy"
                  src=""
                  style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent; z-index: 1; opacity: 1;"
                />
              </div>
              <div
                class="mb-0 text-center card-body"
              >
                <div
                  class="mpccard-name card-subtitle h6"
                >
                  Card 1
                </div>
                <div
                  class="mpccard-spacing"
                >
                  <p
                    class="mpccard-source card-text"
                  >
                    Source 1 [1200 DPI]
                  </p>
                </div>
              </div>
            </div>
          </div>
          <div
            class="mpccard mpccard-hover card"
          >
            <div
              class="pb-0 text-center card-header"
            >
              <p
                class="mpccard-slot"
              >
                Option 2
              </p>
            </div>
            <div>
              <div
                class="rounded-lg shadow-lg ratio ratio-7x5"
                style="z-index: 0;"
              >
                <img
                  alt="Card 2"
                  class="card-img card-img-fade-in"
                  data-nimg="fill"
                  decoding="async"
                  loading="lazy"
                  src=""
                  style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent; z-index: 1; opacity: 1;"
                />
              </div>
              <div
                class="mb-0 text-center card-body"
              >
                <div
                  class="mpccard-name card-subtitle h6"
                >
                  Card 2
                </div>
                <div
                  class="mpccard-spacing"
                >
                  <p
                    class="mpccard-source card-text"
                  >
                    Source 1 [1200 DPI]
                  </p>
                </div>
              </div>
            </div>
          </div>
          <div
            class="mpccard mpccard-hover card"
          >
            <div
              class="pb-0 text-center card-header"
            >
              <p
                class="mpccard-slot"
              >
                Option 3
              </p>
            </div>
            <div>
              <div
                class="rounded-lg shadow-lg ratio ratio-7x5"
                style="z-index: 0;"
              >
                <img
                  alt="Card 3"
                  class="card-img card-img-fade-in"
                  data-nimg="fill"
                  decoding="async"
                  loading="lazy"
                  src=""
                  style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent; z-index: 1; opacity: 1;"
                />
              </div>
              <div
                class="mb-0 text-center card-body"
              >
                <div
                  class="mpccard-name card-subtitle h6"
                >
                  Card 3
                </div>
                <div
                  class="mpccard-spacing"
                >
                  <p
                    class="mpccard-source card-text"
                  >
                    Source 1 [1200 DPI]
                  </p>
                </div>
              </div>
            </div>
          </div>
          <div
            class="mpccard mpccard-hover card"
          >
            <div
              class="pb-0 text-center card-header"
            >
              <p
                class="mpccard-slot"
              >
                Option 4
              </p>
            </div>
            <div>
              <div
                class="rounded-lg shadow-lg ratio ratio-7x5"
                style="z-index: 0;"
              >
                <img
                  alt="Card 4"
                  class="card-img card-img-fade-in"
                  data-nimg="fill"
                  decoding="async"
                  loading="lazy"
                  src=""
                  style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent; z-index: 1; opacity: 1;"
                />
              </div>
              <div
                class="mb-0 text-center card-body"
              >
                <div
                  class="mpccard-name card-subtitle h6"
                >
                  Card 4
                </div>
                <div
                  class="mpccard-spacing"
                >
                  <p
                    class="mpccard-source card-text"
                  >
                    Source 1 [1200 DPI]
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="modal-footer"
    >
      <button
        class="btn btn-secondary"
        type="button"
      >
        Close
      </button>
    </div>
  </div>
</div>
`;

exports[`the html structure of a CardSlot's grid selector, cards grouped together 1`] = `
<div
  class="modal-dialog modal-lg"
  data-testid="front-slot0-grid-selector"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="modal-title h4"
      >
        Select Version
      </div>
      <button
        aria-label="Close"
        class="btn-close"
        type="button"
      />
    </div>
    <div
      class="modal-body"
    >
      <div
        class="row"
      >
        <div
          class="col-md-8 col-sm-6"
        >
          <div
            class="btn toggle btn-md off btn-info"
            role="button"
            style="width: 100%; height: 38px;"
          >
            <div
              class="toggle-group"
            >
              <span
                class="btn toggle-on btn-md flex-centre btn-success"
              >
                Facet By Source
              </span>
              <span
                class="btn toggle-off btn-md flex-centre btn-info"
              >
                Group All Cards Together
              </span>
              <span
                class="toggle-handle btn btn-md btn-default"
              />
            </div>
          </div>
        </div>
      </div>
      <br />
      <div
        class="g-0 row row-cols-xxl-4 row-cols-xl-4 row-cols-lg-3 row-cols-md-2 row-cols-sm-2 row-cols-2"
      >
        <div
          class="mpccard mpccard-hover card"
        >
          <div
            class="pb-0 text-center card-header"
          >
            <p
              class="mpccard-slot"
            >
              Option 1
            </p>
          </div>
          <div>
            <div
              class="rounded-lg shadow-lg ratio ratio-7x5"
              style="z-index: 0;"
            >
              <img
                alt="Card 1"
                class="card-img card-img-fade-in"
                data-nimg="fill"
                decoding="async"
                loading="lazy"
                src=""
                style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent; z-index: 1; opacity: 1;"
              />
            </div>
            <div
              class="mb-0 text-center card-body"
            >
              <div
                class="mpccard-name card-subtitle h6"
              >
                Card 1
              </div>
              <div
                class="mpccard-spacing"
              >
                <p
                  class="mpccard-source card-text"
                >
                  Source 1 [1200 DPI]
                </p>
              </div>
            </div>
          </div>
        </div>
        <div
          class="mpccard mpccard-hover card"
        >
          <div
            class="pb-0 text-center card-header"
          >
            <p
              class="mpccard-slot"
            >
              Option 2
            </p>
          </div>
          <div>
            <div
              class="rounded-lg shadow-lg ratio ratio-7x5"
              style="z-index: 0;"
            >
              <img
                alt="Card 2"
                class="card-img card-img-fade-in"
                data-nimg="fill"
                decoding="async"
                loading="lazy"
                src=""
                style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent; z-index: 1; opacity: 1;"
              />
            </div>
            <div
              class="mb-0 text-center card-body"
            >
              <div
                class="mpccard-name card-subtitle h6"
              >
                Card 2
              </div>
              <div
                class="mpccard-spacing"
              >
                <p
                  class="mpccard-source card-text"
                >
                  Source 1 [1200 DPI]
                </p>
              </div>
            </div>
          </div>
        </div>
        <div
          class="mpccard mpccard-hover card"
        >
          <div
            class="pb-0 text-center card-header"
          >
            <p
              class="mpccard-slot"
            >
              Option 3
            </p>
          </div>
          <div>
            <div
              class="rounded-lg shadow-lg ratio ratio-7x5"
              style="z-index: 0;"
            >
              <img
                alt="Card 3"
                class="card-img card-img-fade-in"
                data-nimg="fill"
                decoding="async"
                loading="lazy"
                src=""
                style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent; z-index: 1; opacity: 1;"
              />
            </div>
            <div
              class="mb-0 text-center card-body"
            >
              <div
                class="mpccard-name card-subtitle h6"
              >
                Card 3
              </div>
              <div
                class="mpccard-spacing"
              >
                <p
                  class="mpccard-source card-text"
                >
                  Source 1 [1200 DPI]
                </p>
              </div>
            </div>
          </div>
        </div>
        <div
          class="mpccard mpccard-hover card"
        >
          <div
            class="pb-0 text-center card-header"
          >
            <p
              class="mpccard-slot"
            >
              Option 4
            </p>
          </div>
          <div>
            <div
              class="rounded-lg shadow-lg ratio ratio-7x5"
              style="z-index: 0;"
            >
              <img
                alt="Card 4"
                class="card-img card-img-fade-in"
                data-nimg="fill"
                decoding="async"
                loading="lazy"
                src=""
                style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent; z-index: 1; opacity: 1;"
              />
            </div>
            <div
              class="mb-0 text-center card-body"
            >
              <div
                class="mpccard-name card-subtitle h6"
              >
                Card 4
              </div>
              <div
                class="mpccard-spacing"
              >
                <p
                  class="mpccard-source card-text"
                >
                  Source 1 [1200 DPI]
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="modal-footer"
    >
      <button
        class="btn btn-secondary"
        type="button"
      >
        Close
      </button>
    </div>
  </div>
</div>
`;
